lib/std/crypto.zig

```diff
@@ -1,57 +1,72 @@
-pub const Md5 = @import("crypto/md5.zig").Md5;
-pub const Sha1 = @import("crypto/sha1.zig").Sha1;
-
-const sha2 = @import("crypto/sha2.zig");
-pub const Sha224 = sha2.Sha224;
-pub const Sha256 = sha2.Sha256;
-pub const Sha384 = sha2.Sha384;
-pub const Sha512 = sha2.Sha512;
-
-const sha3 = @import("crypto/sha3.zig");
-pub const Sha3_224 = sha3.Sha3_224;
-pub const Sha3_256 = sha3.Sha3_256;
-pub const Sha3_384 = sha3.Sha3_384;
-pub const Sha3_512 = sha3.Sha3_512;
-
-pub const gimli = @import("crypto/gimli.zig");
-
-const blake2 = @import("crypto/blake2.zig");
-pub const Blake2s224 = blake2.Blake2s224;
-pub const Blake2s256 = blake2.Blake2s256;
-pub const Blake2b384 = blake2.Blake2b384;
-pub const Blake2b512 = blake2.Blake2b512;
-
-pub const Blake3 = @import("crypto/blake3.zig").Blake3;
-
-const hmac = @import("crypto/hmac.zig");
-pub const HmacMd5 = hmac.HmacMd5;
-pub const HmacSha1 = hmac.HmacSha1;
-pub const HmacSha256 = hmac.HmacSha256;
-pub const HmacBlake2s256 = hmac.HmacBlake2s256;
-
-pub const chacha20 = @import("crypto/chacha20.zig");
-pub const chaCha20IETF = chacha20.chaCha20IETF;
-pub const chaCha20With64BitNonce = chacha20.chaCha20With64BitNonce;
-pub const xChaCha20IETF = chacha20.xChaCha20IETF;
-
-pub const Poly1305 = @import("crypto/poly1305.zig").Poly1305;
-
-const import_aes = @import("crypto/aes.zig");
-pub const AES128 = import_aes.AES128;
-pub const AES256 = import_aes.AES256;
-
-pub const Curve25519 = @import("crypto/25519/curve25519.zig").Curve25519;
-pub const Ed25519 = @import("crypto/25519/ed25519.zig").Ed25519;
-pub const Edwards25519 = @import("crypto/25519/edwards25519.zig").Edwards25519;
-pub const X25519 = @import("crypto/25519/x25519.zig").X25519;
-pub const Ristretto255 = @import("crypto/25519/ristretto255.zig").Ristretto255;
+/// Hash functions.
+pub const hash = struct {
+    /// Legacy hash functions, that shouldn't be used in new applications.
+    pub const legacy = struct {
```

Comment:
The API of everything in the std lib should be descriptive rather than prescriptive. Rather than implying "don't use this API because it is legacy" it should instead describe clearly what the API is and what it does, and let the programmer make the choice.

MD5 and SHA1, while being poor choices for new applications and libraries to choose, are well-defined mathematical descriptions of data. There are plenty of use cases for these APIs that would not fall under the "legacy" category. As an example off the top of my head, the DWARFv4 specification uses MD5 as a way for applications to annotate the "version" of a source file so that debuggers can tell if it has changed. Sure, it would have been nice if they chose a different cryptographic algorithm, but they didn't. And many debugging tools only support up to DWARFv4. So even the (brand new!) self-hosted zig compiler code has a use case for MD5 and "legacy" has nothing to do with it - it's part of the ABI that we are limited to based on what other downstream tools support.

The preferred way to discourage programmers from using these APIs is strongly worded doc comments. But even here, not by telling people what to do, but by laying out the facts and letting the programmer make the decision. For example, I might choose to write something like this:

```
/// MD5 is hash function producing a 128-bit value that is cryptographically broken and therefore unsuitable for cryptographic use.
/// For a fast-performing, cryptographically secure hash function, see <insert the best recommendation here>.
```

If you wanted to break the hash namespace into `secure` and `broken` I could be convinced of that.

Reply:
Got it!

They have been moved to the same namespace as other functions, and comments have been added.

Thanks everyone for your feedback!